saw-core-coq/src/Verifier/SAW/Translation/Coq/Term.hs

```diff
@@ -425,9 +428,13 @@ withTopTranslationState m =
      modify $ set nextSharedName "var__0"
      m
 
-mkDefinition :: Coq.Ident -> Coq.Term -> Coq.Decl
-mkDefinition name (Coq.Lambda bs t) = Coq.Definition name bs Nothing t
-mkDefinition name t = Coq.Definition name [] Nothing t
+-- | Generate a Coq @Definition@ with a given name, body, and type, using the
+-- lambda-bound variable names for the variables if they are available
+mkDefinition :: Coq.Ident -> Coq.Term -> Coq.Term -> Coq.Decl
+mkDefinition name (Coq.Lambda bs t) (Coq.Pi bs' tp)
+  | length bs' == length bs =
```

Comment:
This `length bs' == length bs` check is rather strange. Are there ever circumstances under which these _won't_ be the same length?

Reply:
It's possible, though not likely, which is why the check is there. One way is that the type assigned to a term could be definitionally but not syntactically equal to a function type. E.g., a type like `lrtToType (LRT_Fun a (\ (x:a) -> LRT_Ret b))`, which is generated by some of the Heapster code, evaluates to the type `a -> b` but is not syntactically equal to a function type. Or it could be that the body being defined is a partially applied function, which would have no lambdas but would have a function type.

If we really want to make sure to add named arguments to a `Definition` in these cases, we could make this functionality smarter, but for now I think it makes sense to just handle the common and easy case.

Reply:
Ah, I see. In that case, I'm fine with leaving this check as-is, perhaps with an accompanying comment to explain what it does and what happens should it fail.